packages/api-fetch/src/middlewares/fetch-all-middleware.js

```diff
@@ -0,0 +1,103 @@
+/* eslint-disable no-console */
+/**
+ * WordPress dependencies
+ */
+import { addQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import apiFetch from '../';
+
+// Apply query arguments to both URL and Path, whichever is present.
+const modifyQuery = ( { path, url, ...options }, queryArgs ) => ( {
+	...options,
+	...( url ? {
+		url: addQueryArgs( url || path, queryArgs ),
+	} : {} ),
+	...( path ? {
+		path: addQueryArgs( path || path, queryArgs ),
+	} : {} ),
+} );
+
+// Duplicates parsing functionality from apiFetch.
+const parseResponse = ( response ) => response.json ?
+	response.json() :
+	Promise.reject( response );
+
+const parseLinkHeader = ( linkHeader ) => {
+	if ( ! linkHeader ) {
+		return {};
+	}
+	const match = linkHeader.match( /<([^>]+)>; rel="next"/ );
+	return match ? {
+		next: match[ 1 ],
+	} : {};
+};
+
+const getNextPageUrl = ( response ) => {
+	const { next } = parseLinkHeader( response.headers.get( 'link' ) );
+	return next;
+};
+
+const requestContainsUnboundedQuery = ( options ) => {
+	const pathIsUnbounded = options.path && options.path.indexOf( 'per_page=-1' > -1 );
+	const urlIsUnbounded = options.url && options.url.indexOf( 'per_page=-1' > -1 );
+	return pathIsUnbounded || urlIsUnbounded;
+};
+
+// The REST API enforces an upper limit on the per_page option. To handle large
+// collections, apiFetch consumers can pass `per_page=-1`; this middleware will
+// then recursively assemble a full response array from all available pages.
+const fetchAllMiddleware = async ( options, next ) => {
+	if ( options.parse === false ) {
+		// If a consumer has opted out of parsing, do not apply middleware.
+		return next( options );
+	}
+	if ( ! requestContainsUnboundedQuery( options ) ) {
+		// If neither url nor path is requesting all items, do not apply middleware.
+		return next( options );
+	}
+
+	// Retrieve requested page of results.
+	const response = await next( {
+		...modifyQuery( options, {
+			per_page: 100,
+		} ),
+		// Ensure headers are returned for page 1.
+		parse: false,
+	} );
+
+	const results = await parseResponse( response );
+
+	if ( ! Array.isArray( results ) ) {
+		// We have no reliable way of merging non-array results.
+		return results;
```

Comment:
I don't think this is a good solution. As we changed the intent of the request but didn't took it into consideration when providing the result.

Reply:
What would you suggest doing in this case? Endpoints which return an object will almost never return a `next` link header anyway, so there's not a huge impact to this that I can see.

Reply:
> Endpoints which return an object will almost never return a next link header anyway

It's the "almost" that's worrying me why can't we just merge the objects?

Reply:
I have not seen an endpoint following our REST API's conventions which would both expose pagination headers and return an object, so I don't know how I should begin to attempt to merge object responses. If we assume that each subsequent page should overwrite the values from the prior, there may be data loss. If we assume that we should only merge array children, that could work, but again, there could be data loss.

Given that we do not have an immediate need to support this for core WordPress endpoints (that I know of), I suggest we look for prior art and revisit the question to add object merging logic once we encounter a specific situation where this becomes an issue.

Reply:
> I have not seen an endpoint following our REST API's conventions which would both expose pagination headers and return an object,

Right now in Gutenberg we pass  `per_page=-1` in a request to fetch taxonomies. This might be wrong though?

Reply:
Not wrong so much as unneedd -- the taxonomies controller does not accept a `per_page` argument, so that query parameter gets ignored. We use `get_object_taxonomies` and `get_taxonomies` within [`wp-rest-taxonomies-controller.php`](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/endpoints/class-wp-rest-taxonomies-controller.php), niether of which accepts a limiting argument.